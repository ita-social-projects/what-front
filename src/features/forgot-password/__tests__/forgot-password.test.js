import React from 'react';
import { useSelector } from 'react-redux';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { useActions } from '@/shared/index.js';
import { forgotPassword } from '@models/index.js';
import { ForgotPassword } from '../forgot-password.js';

jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockReturnValue({
    isLoading: false,
    error: '',
    loaded: true,
  }),
}));

jest.mock('@/shared/hooks', () => ({
  useActions: jest.fn().mockReturnValue([jest.fn()]),
}));

describe('Forgot password', () => {
  let formValues;
  beforeEach(() => {
    formValues = {
      email: 'testmail@gmail.com',
    };
  });

  it('should render component', () => {
    const { container } = render(
      <Router>
        <ForgotPassword />
      </Router>
    );
    const forgotPasswordContainer = container.getElementsByClassName(
      'container'
    );

    expect(forgotPasswordContainer).toMatchSnapshot();
  });

  it('button should be disabled', async () => {
    const { getByText } = render(
      <Router>
        <ForgotPassword />
      </Router>
    );

    const submitButton = getByText('Send');

    expect(submitButton).toBeDisabled();
  });

  it('should handle error receiving', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      error: 'Something went wrong',
      loaded: true,
    });

    const { getByText } = render(
      <Router>
        <ForgotPassword />
      </Router>
    );

    const error = getByText('Something went wrong');
    expect(error).toBeInTheDocument();
  });

  it('should submit the form with correct schema', async () => {
    const { getByPlaceholderText, getByText } = render(
      <Router>
        <ForgotPassword />
      </Router>
    );

    const email = getByPlaceholderText('Email address');
    const submitButton = getByText('Send');

    const [setForgotPassword] = useActions(forgotPassword);

    fireEvent.change(email, {
      target: {
        value: formValues.email,
      },
    });

    await waitFor(() => {
      fireEvent.click(submitButton);
    });

    setForgotPassword({
      email: email.value,
    });

    expect(setForgotPassword).toHaveBeenCalledWith(formValues);
  });
});
