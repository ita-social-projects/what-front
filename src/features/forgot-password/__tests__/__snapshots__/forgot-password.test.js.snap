// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot password should render component 1`] = `
HTMLCollection [
  <div
    class="container"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-xl-6 col-md-8 col-sm-10 col-12"
      >
        <div
          class="card, shadow"
        >
          <form
            action="#"
            class="p-3"
            novalidate=""
          >
            <h3
              class="text-center"
            >
              Forgot your password?
            </h3>
            <p
              class="text-center mt-3"
            >
              Don't worry! Just fill in your email and we'll send you a link to reset your password
            </p>
            <hr />
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email address
              </label>
              <input
                class="form-control"
                id="email"
                name="email"
                placeholder="Email address"
                type="email"
                value=""
              />
            </div>
            <div
              class="d-flex justify-content-center"
            >
              <button
                class="btn undefined"
                disabled=""
                type="submit"
              >
                Send
              </button>
            </div>
            
          </form>
        </div>
      </div>
    </div>
  </div>,
]
`;
